### PR TITLE
Allow to use 0 to n separate component labels

### PR DESCRIPTION
**Pull request by Luthaf**
This PR makes the component labels a list of `Labels` instead of a single set of `Labels`. This has a handful of advantages:

- invariants are now represented with a real 2D array, and not a `(n_samples, 1, n_features)` array
- we can have extra component dimensions for gradients: gradients of lambda SOAP would look like this: `(structure, atom) / gradient_direction / spherical_mu / (n1, n2, l)`
- data with multiple components are cleanly separated (for example uncoupled hamiltonian can be stored with two separate mu labels) & easier to access `hamiltonian[sample, mu_1, mu_2, ..]`

----

@ceriottm @picoCentauri @curiosity54 this will break all of the examples since the Python API changes. The two main changes are (1) `block.components` is now a list of labels instead of a single label; (2) getting a gradient returns a proxy object which can be used to access the actual data.

```py
# before
gradient_samples, gradient_data = block.gradient("positions")
# components & features are the same ones as the values
gradient_components = block.components
gradient_features = block.features

# after
gradient = block.gradient("positions")

gradient_samples = gradient.samples
gradient_components = gradient.components
# feature labels are still the same as the values'
gradient_features = gradient.features
gradient_data = gradient.data
```

Let me know if you have any feedback on this!

### CODE REVIEW

File: aml_storage/block.py [hunk 1]
```diff
@@ -1,33 +1,40 @@
...
-from .data import AmlData, Array, aml_array_t_to_python_object
+from .data import AmlData, Array, aml_array_to_python_object
 
 
 class Block:
     """
-    Basic building block for descriptor. A single block contains a 3-dimensional
-    :py:class:`aml_storage.data.Array`, and three sets of :py:class:`Labels`
-    (one for each dimension).
+    Basic building block for descriptor. A single block contains a n-dimensional
```
Comment (PicoCentauri) [inline aml_storage/block.py:17]: I not sure how strictly we would like to follow [PEP257](https://peps.python.org/pep-0257/#multi-line-docstrings) but I would stick with a first line short description and then a new paragraph.

```suggestion
    Basic building block for descriptor. 
    
    A single block contains a n-dimensional
```

File: aml_storage/block.py [hunk 2]
```diff
@@ -1,33 +1,40 @@
...
+    :py:class:`aml_storage.data.Array`, and n sets of :py:class:`Labels` (one
+    for each dimension). The first dimension is the *samples* dimension, the
+    last dimension is the *features* dimension, and any intermediary dimension
```
Comment (PicoCentauri) [inline aml_storage/block.py:20]:
```suggestion
    last dimension is the *features* dimension. Any intermediary dimension
```

File: aml_storage/block.py [hunk 3]
```diff
@@ -116,82 +128,104 @@ def copy(self) -> "Block":
...
-    def _labels(self, name: str, kind) -> Labels:
-        """Get the labels of the given ``kind`` for the ``name`` array in this
-        block"""
-        result = aml_labels_t()
-        self._lib.aml_block_labels(self._ptr, name.encode("utf8"), kind.value, result)
-        return Labels._from_aml_labels_t(result, parent=self)
+    @property
+    def values(self) -> Array:
+        """
+        Access the values for this block. The array type depends on how the
```
Comment (PicoCentauri) [inline aml_storage/block.py:134]:
```suggestion
        Access the values for this block.
        
        The array type depends on how the
```

File: aml_storage/block.py [hunk 4]
```diff
@@ -116,82 +128,104 @@ def copy(self) -> "Block":
...
         these labels describe the first dimension of the ``values`` array.
         """
-        return self._labels("values", aml_label_kind.AML_SAMPLE_LABELS)
+        return self._labels(0)
 
     @property
-    def components(self) -> Labels:
+    def components(self) -> List[Labels]:
         """
         Access the component :py:class:`Labels` for this block. The entries in
```
Comment (PicoCentauri) [inline aml_storage/block.py:153]:
```suggestion
        Access the component :py:class:`Labels` for this block.
        
        The entries in
```

File: aml_storage/block.py [hunk 5]
```diff
@@ -116,82 +128,104 @@ def copy(self) -> "Block":
...
+        result = []
+        for axis in range(n_components):
+            result.append(self._labels(axis + 1))
+
+        return result
 
     @property
     def features(self) -> Labels:
         """
         Access the feature :py:class:`Labels` for this block. The entries in
```
Comment (PicoCentauri) [inline aml_storage/block.py:167]:
```suggestion
        Access the feature :py:class:`Labels` for this block.
        
        The entries in
```

File: aml_storage/block.py [hunk 6]
```diff
@@ -116,82 +128,104 @@ def copy(self) -> "Block":
...
-            gradient._storage,
+            data.aml_array,
+            samples._as_aml_labels_t(),
+            components_array,
+            len(components_array),
         )
 
     def gradients_list(self) -> List[str]:
-        """ """
+        """Get a list of all gradients defined in this block"""
```
Comment (PicoCentauri) [inline aml_storage/block.py:228]:
```suggestion
        """Get a list of all gradients defined in this block."""
```

File: aml_storage/block.py [hunk 7]
```diff
@@ -211,10 +245,79 @@ def has_gradient(self, parameter: str) -> bool:
         """
         return parameter in self.gradients_list()
 
-    def _get_array(self, name: str) -> Array:
-        return aml_array_t_to_python_object(self._get_raw_array(name)).array
+    def gradients(self) -> Generator[Tuple[str, "Gradient"], None, None]:
+        """
+        Get an iterator over all gradients defined in this block
+        """
```
Comment (PicoCentauri) [inline aml_storage/block.py:251]:
```suggestion
        """Get an iterator over all gradients defined in this block."""
```

File: aml_storage/block.py [hunk 8]
```diff
@@ -211,10 +245,79 @@ def has_gradient(self, parameter: str) -> bool:
...
+
+        self._block = block
+        self._name = name
+
+    @property
+    def data(self) -> Array:
+        """
+        Access the data for this gradient. The array type depends on how the
+        block was created. Currently, numpy ``ndarray`` and torch ``Tensor`` are
+        supported.
```
Comment (PicoCentauri) [inline aml_storage/block.py:273]:
```suggestion
        Access the data for this gradient.
        
        The array type depends on how the block was created.
        Currently, numpy ``ndarray`` and torch ``Tensor`` are
        supported.
```

File: aml_storage/block.py [hunk 9]
```diff
@@ -211,10 +245,79 @@ def has_gradient(self, parameter: str) -> bool:
...
+
+        raw_array = _get_raw_array(self._lib, self._block._ptr, self._name)
+        return aml_array_to_python_object(raw_array).array
+
+    @property
+    def samples(self) -> Labels:
+        """
+        Access the sample :py:class:`Labels` for this gradient. The entries in
+        these labels describe the first dimension of the ``data`` array.
+        """
```
Comment (PicoCentauri) [inline aml_storage/block.py:284]:
```suggestion
        """
        Access the sample :py:class:`Labels` for this gradient.
        
        The entries in these labels describe the first dimension of
        the ``data`` array.
        """
```

File: aml_storage/block.py [hunk 10]
```diff
@@ -211,10 +245,79 @@ def has_gradient(self, parameter: str) -> bool:
...
+        return self._labels(0)
+
+    @property
+    def components(self) -> List[Labels]:
+        """
+        Access the component :py:class:`Labels` for this gradient. The entries
+        in these labels describe intermediate dimensions of the ``data`` array.
+        """
```
Comment (PicoCentauri) [inline aml_storage/block.py:292]:
```suggestion
        """
        Access the component :py:class:`Labels` for this gradient.
        
        The entries in these labels describe intermediate dimensions of
        the ``data`` array.
        """
```

File: aml_storage/block.py [hunk 11]
```diff
@@ -211,10 +245,79 @@ def has_gradient(self, parameter: str) -> bool:
...
+        return result
+
+    @property
+    def features(self) -> Labels:
+        """
+        Access the feature :py:class:`Labels` for this gradient. The entries in
+        these labels describe the last dimension of the ``data`` array. The
+        features are guaranteed to be the same for values and gradients in the
+        same block.
+        """
```
Comment (PicoCentauri) [inline aml_storage/block.py:308]:
```suggestion
        """
        Access the feature :py:class:`Labels` for this gradient.
        
        The entries in these labels describe the last dimension of 
        the ``data`` array. The features are guaranteed to be the same
        for values and gradients in the
        same block.
        """
```

**Review by PicoCentauri — COMMENTED** [review]
I just have some minor doc comments.

Should we create an PR the examples repo and merge both simultaneously?

File: aml_storage/block.py (hunk 1 above)
Comment (Luthaf) [reply aml_storage/block.py:17]: Thanks a lot for the review! 

I have an upcoming PR which renames a bunch of types/properties, I'll add the doc changes to this other PR.

**Comment by Luthaf** [comment]
> Should we create an PR the examples repo and merge both simultaneously?

I would merge this and #14, and then make all the required changes to examples at the same time. Hopefully all the changes should be pretty simple to do.

**Comment by Luthaf** [comment]
Also, this and #14 might not be merged immediately, but after we are relatively sure that's the best level of abstraction for what we try to do here.

File: aml_storage/block.py (hunk 1 above)
Comment (PicoCentauri) [reply aml_storage/block.py:17]: Totally fine with this. If you want, I can also look at the documentation afterwards? We could add links to numpy and torch with intersphinx.

Comment (Luthaf) [reply aml_storage/block.py:17]: another set of eyes on the documentation would be great!

**Pull request merged**